src/main/kotlin/org/rust/lang/core/ControlFlowGraph.kt

```diff
@@ -0,0 +1,199 @@
+        }
+
+        graph.forEachNode { node ->
+            val element = node.data.element
+            if (element != null) table.getOrPut(element, ::mutableListOf).add(node)
+        }
+
+        return table
+    }
+
+    fun depthFirstTraversalTrace(): String =
+        graph.depthFirstTraversal(this.entry).joinToString("\n") { it.data.text }
+
+
+    /**
+     * Creates graph description written in the DOT language.
+     * Usage: copy the output into `cfg.dot` file and run `dot -Tpng cfg.dot -o cfg.png`
+     */
+    fun createDotDescription(): String =
+        buildString {
+            append("digraph {\n")
+
+            graph.forEachEdge { edge ->
+                val source = edge.source
+                val target = edge.target
+                val sourceNode = source.data
+                val targetNode = target.data
+
+                append("    \"${source.index}: ${sourceNode.text}\" -> \"${target.index}: ${targetNode.text}\";\n")
+            }
+
+            append("}\n")
+        }
+}
+
+
+sealed class ExitPoint {
+    class Return(val e: RsRetExpr) : ExitPoint()
+    class TryExpr(val e: RsExpr) : ExitPoint() // `?` or `try!`
+    class DivergingExpr(val e: RsExpr) : ExitPoint()
+    class TailExpr(val e: RsExpr) : ExitPoint()
+    class TailStatement(val stmt: RsExprStmt) : ExitPoint()
+
+    companion object {
+        fun process(fn: RsFunction, sink: (ExitPoint) -> Unit) {
+            fn.block?.acceptChildren(ExitPointVisitor(sink))
+        }
+
+        fun process(fn: RsLambdaExpr, sink: (ExitPoint) -> Unit) {
+            fn.expr?.acceptChildren(ExitPointVisitor(sink))
+        }
+    }
+}
+
+private class ExitPointVisitor(
+    private val sink: (ExitPoint) -> Unit
+) : RsVisitor() {
+    override fun visitElement(element: RsElement) = element.acceptChildren(this)
+
+    override fun visitLambdaExpr(lambdaExpr: RsLambdaExpr) = Unit
+    override fun visitFunction(function: RsFunction) = Unit
+
+    override fun visitRetExpr(retExpr: RsRetExpr) = sink(ExitPoint.Return(retExpr))
+
+    override fun visitTryExpr(tryExpr: RsTryExpr) {
+        tryExpr.expr.acceptChildren(this)
+        sink(ExitPoint.TryExpr(tryExpr))
+    }
+
+    override fun visitMacroExpr(macroExpr: RsMacroExpr) {
+        if (macroExpr.macroCall.tryMacroArgument != null) sink(ExitPoint.TryExpr(macroExpr))
+        if (macroExpr.type == TyNever) sink(ExitPoint.DivergingExpr(macroExpr))
+    }
+
+    override fun visitExpr(expr: RsExpr) {
+        when (expr) {
+            is RsIfExpr,
+            is RsBlockExpr,
+            is RsMatchExpr -> expr.acceptChildren(this)
+            else -> {
+                if (expr.isInTailPosition) sink(ExitPoint.TailExpr(expr)) else expr.acceptChildren(this)
+            }
+        }
+    }
+
+    override fun visitExprStmt(exprStmt: RsExprStmt) {
+        exprStmt.acceptChildren(this)
+        val block = exprStmt.parent as? RsBlock ?: return
```

Comment:
Maybe you can put those computations in some separate method with descriptive name